docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -173,12 +179,36 @@ With the following validations:
 - `KubeadmControlPlane.Spec.KubeadmConfigSpec` allows mutations required for supporting following use cases:
     - Change of imagesRepository/imageTags (with validation of CoreDNS supported upgrade)
     - Change of node registration options
-    - Change of pre/post kubeadm commands 
+    - Change of pre/post kubeadm commands
     - Change of cloud init files
 
 And the following defaulting:
 
-- `KubeadmControlPlane.Spec.Replicas: 1` 
+- `KubeadmControlPlane.Spec.Replicas: 1`
+
+##### Changes for supporting different upgrade strategies
+
+- Add `KubeadmControlPlane.Spec.UpgradeStrategy` defined as:
+
+```go
+    // UpgradeStrategy is a field to select an upgrade strategy. If field is not present
+    // KCP uses default scale out strategy. Only supported additional upgrade strategy is ScaleIn.
+    // +optional
+    UpgradeStrategy UpgradeStrategyType `json:"upgradeStrategy,omitempty"`
+```
+
+- Add `KubeadmControlPlane.UpgradeStrategyType` struct defined as:
+
+```go
+    // UpgradeStrategyType defines the type of upgrade strategy. This struct is similar to MachineDeploymentStrategy in
+    // https://github.com/kubernetes-sigs/cluster-api/blob/master/api/v1alpha3/machinedeployment_types.go
+    // Struct allows addition of the new upgrade strategies and extensions to existing ones.  
+    type UpgradeStrategyType struct {
```

Comment:
Just to double-check this, we are introducing the nested struct because we do expect in the foreseeable future to add a more field here?

Reply:
Yes, that was my original idea.